moveit_ros/planning/planning_scene_monitor/test/trajectory_monitor_tests.cpp

```diff
@@ -0,0 +1,111 @@
+ *
+ *  Copyright (c) 2021, PickNik, Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Abishalini Sivaraman */
+
+#include <memory>
+#include <chrono>
+#include <thread>
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "moveit/planning_scene_monitor/trajectory_monitor.h"
+#include "moveit/planning_scene_monitor/current_state_monitor.h"
+#include "moveit/utils/robot_model_test_utils.h"
+#include "rclcpp/node_interfaces/node_clock_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "tf2_ros/buffer.h"
+
+using namespace std::chrono_literals;
+
+struct MockTrajectoryMonitorMiddlewareHandle : public planning_scene_monitor::TrajectoryMonitor::MiddlewareHandle
+{
+  MOCK_METHOD(void, sleep, (), (override));
+};
+
+struct MockCurrentStateMonitorMiddlewareHandle : public planning_scene_monitor::CurrentStateMonitor::MiddlewareHandle
+{
+  MOCK_METHOD(rclcpp::Time, now, (), (const, override));
+  MOCK_METHOD(void, createJointStateSubscription,
+              (const std::string& topic, planning_scene_monitor::JointStateUpdateCallback callback), (override));
+  MOCK_METHOD(void, resetJointStateSubscription, (), (override));
+  MOCK_METHOD(std::string, getJointStateTopicName, (), (const, override));
+};
+
+void waitFor(std::chrono::seconds timeout, std::function<bool()> done)
```

Comment:
Doxygen comment would be good.
Does passing `timeout` by `const&` make sense @tylerjw https://stackoverflow.com/questions/45740179/why-arent-stdchronoduration-instances-passed-by-value-in-the-standard-libra or are we going to assume we know it's a POD and there's no point or...

Reply:
I think we assume these are POD and leave it as is.  I didn't know about that... the more you know :rainbow:

Reply:
@griswaldbrooks I love your recent in-depths-backed-up-by-references reviews! They make for great reading! :heart:

If @tylerjw feels it's unnecessary let's leave it as is. My main complain would be the interface itself, because condition variables/void futures should usually be preferred over polling arbitrary condition functions. :cat2: 
But in this scope I think it's actually bikeshedding. :shrug:

Reply:
> If @tylerjw feels it's unnecessary let's leave it as is

Agreed. Especially since this is test code. I feel like test code, while still requiring high quality, can skip on a few things. Not in this case but in others I might write two functions that did something **very** similar with the same name in two different namespaces in the same file.
In production code I would never do such a thing, but I think since the scope of test code tends towards being very local, the value of the implementation being nearby and specific, rather than having to understand different initialization options for example if preferred. Though I might be able to be convinced otherwise.

>  because condition variables/void futures 

I also prefer that but something about it in this scope didn't quite hit right. I think it had something to do with being in the same thread but that sounds funny. The atomic is doing something similar-ish. The function isn't arbitrary since it's the only interface the caller has to know if any samples were collected.

:bicyclist:

Reply:
> They make for great reading! 

I'm trying to prevent myself from picking up habits that aren't based on particular guidance.
I've lived through a few codebases where things were done they way they are because they were done they way they were.